File: Top Interview Questions/src/189.rotate-array.js
/*
 * @lc app=leetcode id=189 lang=javascript
 *
 * [189] Rotate Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
  k %= nums.length;

  var rotateCount = 0;
  var startIdx = 0;
  while (rotateCount < nums.length) {
    var currentIdx = startIdx;
    var prev = nums[currentIdx];
    do {
      currentIdx = (currentIdx + k ) % nums.length;
      
      var temp = nums[currentIdx];
      nums[currentIdx] = prev;
      prev = temp;
      rotateCount++;

    } while (currentIdx !== startIdx);

    startIdx++;
  }

  return nums;
};
// @lc code=end

File: Top Interview Questions/src/189.rotate-array.java
/*
 * @lc app=leetcode id=189 lang=java
 *
 * [189] Rotate Array
 */

// @lc code=start
class Solution {
  public void rotate(int[] nums, int k) {
    k %= nums.length;
  
    int[] answer = new int[nums.length];

    System.arraycopy(nums, nums.length - k, answer, 0, k);
    System.arraycopy(nums, 0, answer, k, nums.length - k);

    System.arraycopy(answer, 0, nums, 0, nums.length);
  }
}
// @lc code=end

